pex/pex_builder.py

```diff
@@ -626,18 +650,141 @@ def freeze(self, bytecode_compile=True):
             self._precompile_source()
         self._frozen = True
 
-    def build(self, filename, bytecode_compile=True, deterministic_timestamp=False):
-        """Package the PEX into a zipfile.
+    def build(
+        self,
+        path,  # type: str
+        bytecode_compile=True,  # type: bool
+        deterministic_timestamp=False,  # type: bool
+        spread=False,  # type: bool
+    ):
+        # type: (...) -> None
+        """Package the PEX application.
+
+        By default, the PEX is packaged as a zipapp for ease of shipping as a single file, but it
+        can also be packaged in spread mode for efficiency of syncing over the network
+        incrementally.
 
-        :param filename: The filename where the PEX should be stored.
+        :param path: The path where the PEX should be stored.
         :param bytecode_compile: If True, precompile .py files into .pyc files.
         :param deterministic_timestamp: If True, will use our hardcoded time for zipfile timestamps.
+        :param spread: If True, will build the PEX in spread directory mode.
 
         If the PEXBuilder is not yet frozen, it will be frozen by ``build``.  This renders the
         PEXBuilder immutable.
         """
         if not self._frozen:
             self.freeze(bytecode_compile=bytecode_compile)
+        if spread:
+            self._build_spreadapp(dirname=path, deterministic_timestamp=deterministic_timestamp)
+        else:
+            self._build_zipapp(filename=path, deterministic_timestamp=deterministic_timestamp)
+
+    def _build_spreadapp(
+        self,
+        dirname,  # type: str
+        deterministic_timestamp=False,  # type: bool
+    ):
+        # type: (...) -> None
+
+        pex_info = self._pex_info
+
+        safe_rmtree(dirname)
+
+        # N.B.: We want an atomic directory, but we don't expect a user to race themselves building
+        # to a single non-PEX_ROOT user-requested output path; so we don't grab an exclusive lock
+        # and dirty the target directory with a `.lck` file.
+        with atomic_directory(dirname, exclusive=False) as atomic_dir:
+            if not atomic_dir.is_finalized:
+                sources = []  # type: List[str]
+                spreads = []  # type: List[Spread]
+
+                work_dir = atomic_dir.work_dir
+
+                # Include user sources, PEX-INFO and __main__ as loose files in src/.
+                for fileset in "source", "resource", "executable", "main", "manifest":
+                    for f in self._chroot.filesets.get(fileset, ()):
+                        dest = os.path.join(work_dir, "src", f)
+                        safe_mkdir(os.path.dirname(dest))
+                        safe_copy(os.path.join(self._chroot.chroot, f), dest)
```

Comment:
It turned out this line was a bug. The `safe_copy` tries to use hard linking, and when that succeeds, you can get a symlink out the other side if `f` went in as a symlink (hardlink of a symlink on linux is a symlink). Wrapping the LHS with an os.path.realpath fixes and lets --spread be used over in Pants where it nets a ~23x space savings vs master when running:
```console
rm -rf ~/.cache/pants/lmdb_store
./pants fmt lint typecheck test build-support:: src:: tests::
``` 
I found:
```console
du -sb lmdb_store.old lmdb_store.main lmdb_store.spread
1623506944	lmdb_store.old
1614954496	lmdb_store.main
1423679488	lmdb_store.spread
```
So that's
old: 1623506944
PEX_PATH transitive: 1614954496 (-8552448) ~8MB saved.
spread: 1423679488 (-199827456) ~200MB saved, ~23x improvement.

IIUC the local lmdb savings should directly map to remote CAS savings / hit ratios.

Reply:
> It turned out this line was a bug. The `safe_copy` tries to use hard linking, and when that succeeds, you can get a symlink out the other side if `f` went in as a symlink (hardlink of a symlink on linux is a symlink). Wrapping the LHS with an os.path.realpath fixes and lets --spread be used over in Pants where it nets a ~23x space savings vs main.

It looks like sandbox outputs are currently symlink aware (because they don't use `PosixFS::new_with_symlink_behavior(..., SymlinkBehavior:Oblivous)` [here](https://github.com/pantsbuild/pants/blob/70f0cd78db860b86c8a611e658e745dffa3f04fc/src/rust/engine/process_execution/src/local.rs#L583-L599)), but that could be disabled if we think that it aligns with expected REAPI behavior? cc @tdyas 

> old: 1623506944
> PEX_PATH transitive: 1614954496 (-8552448) ~8MB saved.
> spread: 1423679488 (-199827456) ~200MB saved, ~23x improvement.

Nice. Did you measure performance while you were at it? Maybe too noisy in larger cases.

Reply:
I did - perf was in the noise. All runs were 17 minutes +/-.

Reply:
> I did - perf was in the noise. All runs were 17 minutes +/-.

Got. Yea, I think that that is likely to dependent on `N` as well (particularly for the pre-`PEX_PATH` change.). Subset building adds up when you do enough of it.